src-java/testing/functional-tests/src/test/groovy/org/openkilda/functionaltests/spec/flows/QinQFlowSpec.groovy

```diff
@@ -967,4 +938,38 @@ class QinQFlowSpec extends HealthCheckSpecification {
             northbound.synchronizeSwitch(swId, true)
         }
     }
+
+    @Memoized
+    List<SwitchPair> getUniqueSwitchPairs(Closure additionalConditions = { true }) {
+        def unpickedUniqueSwitches = topology.activeSwitches.collect { it.hwSwString }.unique(false)
+        def suitablePairs = topologyHelper.getSwitchPairs(true).findAll {
+            additionalConditions(it)
+        }
+        def unpickedSuitableSwitches = unpickedUniqueSwitches.intersect(
+                suitablePairs.collectMany {[it.src.hwSwString, it.dst.hwSwString]}.unique(false))
+        def untestedSwitches = unpickedUniqueSwitches - unpickedSuitableSwitches
+        if(untestedSwitches) {
+            log.warn("Switches left untested: ${untestedSwitches.inspect()}")
+        }
+        assumeFalse(unpickedSuitableSwitches.empty, "No switches that match required conditions") //not possible?
+        def result = []
+        while (!unpickedSuitableSwitches.empty) {
+            def pairs = suitablePairs.sort(false) { swPair ->
+                def score = 0
+                swPair.src.hwSwString in unpickedSuitableSwitches && score++
+                swPair.dst.hwSwString in unpickedSuitableSwitches && score++
+                swPair.src.traffGens && swPair.dst.traffGens && score++
+                return score
+            }
+            //pick a highest score pair, update list of unpicked switches, re-run
+            def pair = pairs.last()
+            result << pair
+            unpickedSuitableSwitches = unpickedSuitableSwitches - pair.src.hwSwString - pair.dst.hwSwString
+        }
+        return result
+    }
+
+    def setupSpec() {
```

Comment:
let's move it to the beginning of the spec.
as for me, it will be more readable. I mean it is easy to understand that we have some code which is before the first feature method is invoked.